comp110-worksheet-6/OxoBoard.cs

```diff
@@ -6,42 +6,168 @@
 		// If there are no empty squares, return true.
 		public bool IsBoardFull()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
-		}
+            bool full;
+            int remaining = 0;
+
+            // Counts number of empty squares
+            for (int y = 0; y < 3; y++)
+            {
+                for (int x = 0; x < 3; x++)
+                {
+                    if (Squares[x, y] == Mark.None)
+                    {
+                        remaining++;
+                    }
+                }
+            }
+
+            // Returns true only if there are no empty squares
+            if (remaining > 0)
+            {
+                full = false;
+            }
+
+            else
+            {
+                full = true;
+            }
+
+            return full;
+        }
 
 		// If a player has three in a row, return Mark.O or Mark.X depending on which player.
 		// Otherwise, return Mark.None.
 		public Mark GetWinner()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            Mark winner = Mark.None;
+
+            // Checks if any X rows are all one type
+            for (int x = 0; x < 3; x++)
+            {
+                int Xrow = 0;
+                int Orow = 0;
+
+                for (int y = 0; y < 3; y++)
+                {
+                    if (Squares[x, y] == Mark.X)
+                    {
+                        Xrow++;
+                    }
+
+                    else if (Squares[x, y] == Mark.O)
+                    {
+                        Orow++;
+                    }
+
+                    if(Xrow == 3)
+                    {
+                        winner = Mark.X;
+                    }
+
+                    if (Orow == 3)
+                    {
+                        winner = Mark.O;
+                    }
+                }
+            }
+
+            // Checks if any Y rows are all one type
+            for (int y = 0; y < 3; y++)
+            {
+                int Xrow = 0;
+                int Orow = 0;
+
+                for (int x = 0; x < 3; x++)
+                {
+                    if (Squares[x, y] == Mark.X)
+                    {
+                        Xrow++;
+                    }
+
+                    else if (Squares[x, y] == Mark.O)
+                    {
+                        Orow++;
+                    }
+
+                    if (Xrow == 3)
+                    {
+                        winner = Mark.X;
+                    }
+
+                    if (Orow == 3)
+                    {
+                        winner = Mark.O;
+                    }
+                }
+            }
+
+
+            // Checks if the down diagonal is all one type
+            if (Squares[0,0] == Squares[1,1] && Squares[0, 0] == Squares[2, 2])
+            {
+                winner = Squares[0, 0];
+            }
```

Comment:
You should also be checking that the squares aren't all `Mark.None` here -- otherwise this code will also detect if the diagonals are all empty